File: frontend/src/main/webapp/app/audienceresearch/audienceresearch.service.ts
import {Injectable} from '@angular/core';
import {Http} from '@angular/http';
import {Router} from '@angular/router';

import {CommonService} from '../common/common.service';
import {AudienceResearch} from "./audienceresearch.model";
import {AudienceResearchStat} from "./audienceresearchstat.model";

@Injectable()
export class AudienceResearchService extends CommonService {

    constructor(public router: Router, public http: Http) {
        super(router, http);
    }

    public getAudienceResearches(): Promise<any> {
        return this.httpGet(this.api.audienceResearch.list, {});
    }

    public getById(id: number): Promise<any> {
        return this.httpGet(this.api.audienceResearch.get, {
            audienceResearchId: id
        });
    }

    public getStat(audienceResearchId: number, channelId: number): Promise<AudienceResearchStat> {
        return this.httpGet(this.api.audienceResearch.stat, {
            audienceResearchId: audienceResearchId,
            channelId: channelId
        });
    }

    public getChannels(text: string, type: string, internalOnly: boolean): Promise<any> {
        return this.httpGet(this.api.audienceResearch.channels, {
            text: text,
            type: type,
            internalOnly: internalOnly
        });
    }

    public getAdvertisers(): Promise<any> {
        return this.httpGet(this.api.audienceResearch.advertisers, {});
    }

    public create(audienceResearch: AudienceResearch): Promise<any> {
        return this.httpPost(this.api.audienceResearch.create, audienceResearch);
    }

    public update(audienceResearch: AudienceResearch): Promise<any> {
        return this.httpPut(this.api.audienceResearch.update, audienceResearch);
    }

    public updateYesterdayComment(comment: string, id: number): Promise<any> {
        return this.httpPut(this.api.audienceResearch.yesterdayComment, null, {
            comment: comment,
            id: id
        });
    }

    public updateTotalComment(comment: string, id: number): Promise<any> {
        return this.httpPut(this.api.audienceResearch.totalComment, null, {
            comment: comment,
            id: id
        });
    }

    public delete(id: number): Promise<any> {
        return this.httpDelete(this.api.audienceResearch.delete, {}, {
            audienceResearchId: id
        });
    }

    public getDynamicLocalizations(channelId: number): Promise<Array<any>>{
        return this.httpGet(this.api.channel.dynamicLocalizations, {
            channelId: channelId
        });
    }
}

File: frontend/src/main/java/app/programmatic/ui/audienceresearch/service/AudienceResearchService.java
package app.programmatic.ui.audienceresearch.service;
import app.programmatic.ui.audienceresearch.dao.model.AudienceResearch;
import app.programmatic.ui.audienceresearch.dao.model.AudienceResearchStat;
import app.programmatic.ui.audienceresearch.validation.ValidateAudienceResearch;
import app.programmatic.ui.audienceresearch.view.AudienceResearchView;
import app.programmatic.ui.channel.dao.model.ChannelEntity;
import app.programmatic.ui.channel.dao.model.ChannelType;

import java.util.List;

import static app.programmatic.ui.common.validation.ValidateMethod.CREATE;
import static app.programmatic.ui.common.validation.ValidateMethod.UPDATE;

public interface AudienceResearchService {

    AudienceResearch findEager(Long id);

    List<AudienceResearchView> findAll();

    List<AudienceResearchView> findForExternal(Long accountId);

    List<ChannelEntity> findChannels(String text, ChannelType type, boolean internalOnly, int maxRows);

    Long create(@ValidateAudienceResearch(CREATE) AudienceResearch audienceResearch);

    Long update(@ValidateAudienceResearch(UPDATE) AudienceResearch audienceResearch);

    Long updateYesterdayComment(String comment, Long id);

    Long updateTotalComment(String comment, Long id);

    Long delete(Long id);

    AudienceResearchStat getStat(Long audienceResearchId, Long channelId);
}
